### PR TITLE
[refactor] Remove PipelineDefinition, ModeDefinition, PresetDefinition

### PR DESCRIPTION
**Pull request by smackesey**
## Summary & Motivation

Internal companion PR: https://github.com/dagster-io/internal/pull/5360

Remove `PipelineDefinition`, `ModeDefinition`, `PresetDefinition`. This PR is the culmination of a workstream that involved downstack transitions of our tests to job/op/graph APIs and removing `@pipeline` and `execute_pipeline`. Upstack in https://github.com/dagster-io/dagster/pull/13497 there are renames of many symbols containing "pipeline" to "job" equivalents.

This is a large PR but many of the changes are difficult to separate because e.g. `mode` is so deeply embedded in the system. Reviewers should be assured by the fact that no snapshots are changed-- I confined this round of changes to the definitions layer, with mode/preset effectively stubbed out in external data structures.

Summary of changes:

- Merge `PipelineDefinition` into `JobDefinition` (update all `PipelineDefinition` references accordingly)
- Refactor `JobDefinition` so that it no longer internally stores `ModeDefinition` and `PresetDefinition`-- instead, store the previous contents of the single mode and preset that a job contained directly on the `JobDefinition`
- Update `JobDefinition` methods that take a `mode` as a parameter to a parameterless form
- Remove a large amount of logic that branched on `JobDefinition` vs `PipelineDefinition`
- Remove logic that branched on mode
- Remove `pipelines` field from repository object, remove parallel "pipeline" methods (e.g. `get_all_pipelines` deleted, all references updated to use (pre-existing) `get_all_jobs`)
- Remove any error handling code for missing modes
- Remove mode as an argument for many functions
- Wherever a mode name needs to be supplied for serialized objects, stub in `DEFAULT_MODE_NAME`
- Delete `ModeDefinition`, `PresetDefinition`, `default_executors`

## How I Tested These Changes

Existing test suite, specifically ensuring no snapshots changed.

### CODE REVIEW

File: python_modules/dagster-graphql/dagster_graphql/implementation/run_config_schema.py [hunk 1]
```diff
@@ -18,25 +20,20 @@
...
 
     check.inst_param(selector, "selector", PipelineSelector)
-    check.opt_str_param(mode, "mode")
 
-    external_pipeline = get_external_pipeline_or_raise(graphene_info, selector)
-
-    if mode is None:
-        mode = external_pipeline.get_default_mode_name()
+    if mode and mode != DEFAULT_MODE_NAME:
+        return GrapheneModeNotFoundError(selector=selector, mode=mode)
```
Comment (dpeng817) [inline python_modules/dagster-graphql/dagster_graphql/implementation/run_config_schema.py:30]: shouldn't be possible anymore right?

File: python_modules/dagster/dagster/_cli/job.py [hunk 2]
```diff
@@ -444,7 +438,7 @@ def execute_launch_command(
     kwargs: Mapping[str, str],
 ) -> DagsterRun:
     preset = cast(Optional[str], kwargs.get("preset"))
-    mode = cast(Optional[str], kwargs.get("mode"))
+    cast(Optional[str], kwargs.get("mode"))
```
Comment (dpeng817) [inline python_modules/dagster/dagster/_cli/job.py:441]: delete

File: python_modules/dagster/dagster/_core/definitions/pipeline_base.py [hunk 3]
```diff
@@ -89,20 +85,20 @@ def _subset_for_execution(
...
         if asset_selection:
             return InMemoryPipeline(
-                self._pipeline_def.get_job_def_for_subset_selection(  # type: ignore  # (must be JobDefinition)
+                self._pipeline_def.get_job_def_for_subset_selection(
                     asset_selection=asset_selection
                 ),
                 asset_selection=asset_selection,
             )
-        if self._pipeline_def.is_subset_pipeline:
+        if self._pipeline_def.is_subset_pipeline:  # should never be called
```
Comment (dpeng817) [inline python_modules/dagster/dagster/_core/definitions/pipeline_base.py:93]: if it should never be called should we just get rid of it?

**Review by dpeng817 — COMMENTED** [review]
Can't believe we're finally here.

Given how sweeping this change is, I think it'll be important to do a frontend pass and just ensure nothing broke functionality wise. When this does end up landing, we should definitely have a dogfood day almost dedicated to it, where we just go through the edge cases of the system and make sure there are no behavior changes.

File: python_modules/dagster/dagster/_cli/job.py [hunk 4]
```diff
@@ -567,88 +549,18 @@ def _create_external_pipeline_run(
...
-        if len(external_pipeline.available_modes) > 1:
-            raise DagsterInvariantViolationError(
-                (
-                    "Pipeline {name} has multiple modes (Available modes: {modes}) and you have "
-                    "attempted to execute it without specifying a mode. Set "
-                    "mode property on the PipelineRun object."
-                ).format(name=external_pipeline.name, modes=external_pipeline.available_modes)
-            )
-        mode = external_pipeline.get_default_mode_name()
-
```
Comment (dpeng817) [inline python_modules/dagster/dagster/_cli/job.py:646]: Todo: review

File: python_modules/dagster/dagster/_core/definitions/job_definition.py [hunk 5]
```diff
@@ -106,193 +141,379 @@ def __init__(
...
-        )
-        self._logger_defs_specified = (
-            _logger_defs_specified
-            if _logger_defs_specified is not None
-            else logger_defs is not None
-        )
-        executor_def = check.opt_inst_param(
-            executor_def, "executor_def", ExecutorDefinition, default=multi_or_in_process_executor
-        )
-        check.opt_mapping_param(
```
Comment (dpeng817) [inline python_modules/dagster/dagster/_core/definitions/job_definition.py:140]: Todo: review

File: python_modules/dagster/dagster/_core/definitions/job_definition.py [hunk 6]
```diff
@@ -653,75 +840,22 @@ def run_request_for_partition(
             partition_key=partition_key,
         )
 
-    @public
-    def with_hooks(self, hook_defs: AbstractSet[HookDefinition]) -> "JobDefinition":
-        """Apply a set of hooks to all op instances within the job."""
-        hook_defs = check.set_param(hook_defs, "hook_defs", of_type=HookDefinition)
```
Comment (dpeng817) [inline python_modules/dagster/dagster/_core/definitions/job_definition.py:659]: why are the `with_` functions being deleted here?

File: python_modules/dagster/dagster/_core/execution/api.py [hunk 7]
```diff
@@ -240,25 +227,14 @@ def execute_run(
...
-                ),
-            )
-        else:
-            # when `execute_run` is directly called, the sub pipeline hasn't been created
-            # note that when we receive the solids to execute via DagsterRun, it won't support
-            # solid selection query syntax
-            pipeline = pipeline.subset_for_execution_from_existing_pipeline(
-                frozenset(dagster_run.solids_to_execute) if dagster_run.solids_to_execute else None,
-                dagster_run.asset_selection,
-            )
```
Comment (dpeng817) [inline python_modules/dagster/dagster/_core/execution/api.py:261]: todo: review

**Review by dpeng817 — COMMENTED** [review]
After a few scans of this, nothing really seems controversial. The fact that no tests change behavior here really is very reassuring as well. 

I'd appreciate @alangenfeld 's trained eye here as well just to make sure I'm not missing anything. I've added a few comments for the places where I think the meat of the changes are, that I would like to take a closer look at as well.

File: python_modules/dagster-graphql/dagster_graphql/implementation/run_config_schema.py (hunk 1 above)
Comment (smackesey) [reply python_modules/dagster-graphql/dagster_graphql/implementation/run_config_schema.py:30]: This is just to keep the behavior of the GQL API constant. Added comment.

File: python_modules/dagster/dagster/_core/definitions/pipeline_base.py (hunk 3 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/pipeline_base.py:93]: That was more of a personal note to myself that I forgot to remove when I was trying to understand the subset system. I'll drop the comment and clean up the logic in a follow-up PR that targets the subsetting system.

File: python_modules/dagster/dagster/_cli/job.py (hunk 4 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/_cli/job.py:646]: Everything here looks fine - just removing code that only pertains to pipelines.

File: python_modules/dagster/dagster/_core/definitions/job_definition.py [hunk 8]
```diff
@@ -106,193 +141,379 @@ def __init__(
...
+
+    @property
+    def dependencies(self) -> DependencyMapping[NodeInvocation]:
+        return self._graph_def.dependencies
 
     @public
     @property
     def executor_def(self) -> ExecutorDefinition:
-        return self.get_mode_definition().executor_defs[0]
+        return self._executor_def or DEFAULT_EXECUTOR_DEF
```
Comment (dpeng817) [inline python_modules/dagster/dagster/_core/definitions/job_definition.py:291]: very nice

File: python_modules/dagster/dagster/_core/definitions/repository_definition/repository_data_builder.py [hunk 9]
```diff
@@ -263,30 +260,22 @@ def build_caching_repository_data_from_list(
...
-            jobs[name] = pipeline_or_job
-        else:
-            pipelines[name] = pipeline_or_job
+    for job_def in jobs.values():
+        job_def.validate_resource_requirements_satisfied()
 
     if default_executor_def:
         for name, job_def in jobs.items():
-            if not job_def._executor_def_specified:  # noqa: SLF001
+            if not job_def.has_specified_executor:
```
Comment (dpeng817) [inline python_modules/dagster/dagster/_core/definitions/repository_definition/repository_data_builder.py:273]: one subtle behavior change here is that after defaults from the definitions object / repo have been applied, `has_specified_executor` will be true on the job definition within the repo, whereas I think it wouldn't before? I guess the only time this would be an issue is if you were for some reason retrieving job off of repo and then re-applying to a different Definitions object? Which I don't see ever happening.

Tl;dr I think it's fine

File: python_modules/dagster/dagster/_core/definitions/job_definition.py (hunk 5 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/job_definition.py:140]: subtle behavior change in that subsequent calls to `with_executor_def` don't preserve whether executor / loggers were originally specified on the job, but realistically I think that's fine.

File: python_modules/dagster/dagster/_core/execution/api.py (hunk 7 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/execution/api.py:261]: same answer here as [1]

**Comment by dpeng817** [comment]
One other note: on the places where instead of mode being removed, DEFAULT_MODE_NAME is specified in place - will downstack PRs be removing these callsites, and the reasoning is limiting scope, or is there actual backcompat need? Assuming the former, if the latter comments would be nice

File: python_modules/dagster/dagster/_core/execution/api.py [hunk 10]
```diff
@@ -141,24 +139,13 @@ def gen_fail_restarted_run_worker():
...
-        pipeline_def = pipeline.get_definition()
-        if isinstance(pipeline_def, PipelineSubsetDefinition):
-            check.invariant(
-                dagster_run.solids_to_execute == pipeline.solids_to_execute,
-                "Cannot execute DagsterRun with solids_to_execute {solids_to_execute} that"
-                " conflicts with pipeline subset {pipeline_solids_to_execute}.".format(
-                    pipeline_solids_to_execute=str_format_set(pipeline.solids_to_execute),  # type: ignore  # (possible none)
-                    solids_to_execute=str_format_set(dagster_run.solids_to_execute),  # type: ignore  # (possible none)
-                ),
-            )
```
[the remark this answers is not in the document]
Comment (yuhan) [reply python_modules/dagster/dagster/_core/execution/api.py:153]: yeah i think that's right. as we're getting rid of `PipelineDefinition`, users don't really have a way to differentiate solids_to_execute from `dagste_run.solids_to_execute`. and i _think_ the unit tests are covering edge cases. that being said, i assumed asset selection would behave the same.

File: python_modules/dagster/dagster/_core/execution/api.py [hunk 11]
```diff
@@ -204,7 +191,7 @@ def execute_run(
     Returns:
         PipelineExecutionResult: The result of the execution.
     """
-    if isinstance(pipeline, PipelineDefinition):
+    if isinstance(pipeline, JobDefinition):
         if isinstance(pipeline, JobDefinition):
```
Comment (yuhan) [inline python_modules/dagster/dagster/_core/execution/api.py:195]: we can simplify this condition right?

Comment (smackesey) [reply python_modules/dagster/dagster/_core/execution/api.py:195]: good catch

**Comment by smackesey** [comment]
>One other note: on the places where instead of mode being removed, DEFAULT_MODE_NAME is specified in place - will downstack PRs be removing these callsites, and the reasoning is limiting scope, or is there actual backcompat need? Assuming the former, if the latter comments would be nice

They'll be removed in upstack PRs. If we need backcompat provision of modes after those removals, it'll be done in the serdes layer using the `old_fields` argument to `whitelist_for_serdes`.

**Review by dpeng817 — APPROVED** [review]
legend status
